contracts/chain-adapters/Arbitrum_CustomGasToken_Adapter.sol

```diff
@@ -0,0 +1,291 @@
+
+    // Amount of gas token allocated to pay for the base submission fee. The base submission fee is a parameter unique to
+    // retryable transactions; the user is charged the base submission fee to cover the storage costs of keeping their
+    // ticket’s calldata in the retry buffer. (current base submission fee is queryable via
+    // ArbRetryableTx.getSubmissionPrice). ArbRetryableTicket precompile interface exists at L2 address
+    // 0x000000000000000000000000000000000000006E.
+    // @dev Unlike in Arbitrum_Adapter, this is immutable because we don't know what precision the custom gas token has.
+    uint256 public immutable L2_MAX_SUBMISSION_COST;
+
+    // L2 Gas price bid for immediate L2 execution attempt (queryable via standard eth*gasPrice RPC)
+    uint256 public constant L2_GAS_PRICE = 5e9; // 5 gWei
+
+    uint256 public constant L2_CALL_VALUE = 0;
+
+    uint32 public constant RELAY_TOKENS_L2_GAS_LIMIT = 300_000;
+    uint32 public constant RELAY_MESSAGE_L2_GAS_LIMIT = 2_000_000;
+
+    // This address on L2 receives extra gas token that is left over after relaying a message via the inbox.
+    address public immutable L2_REFUND_L2_ADDRESS;
+
+    ArbitrumL1InboxLike public immutable L1_INBOX;
+
+    ArbitrumL1ERC20GatewayLike public immutable L1_ERC20_GATEWAY_ROUTER;
+
+    // This token is used to pay for l1 to l2 messages if its configured by an Arbitrum orbit chain.
+    IERC20 public immutable CUSTOM_GAS_TOKEN;
+
+    FunderInterface public immutable CUSTOM_GAS_TOKEN_FUNDER;
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l1ArbitrumInbox Inbox helper contract to send messages to Arbitrum.
+     * @param _l1ERC20GatewayRouter ERC20 gateway router contract to send tokens to Arbitrum.
+     * @param _l2RefundL2Address L2 address to receive gas refunds on after a message is relayed.
+     * @param _l1Usdc USDC address on L1.
+     * @param _cctpTokenMessenger TokenMessenger contract to bridge via CCTP.
+     * @param _customGasToken Custom gas token to pay for L1 to L2 messages.
+     * @param _customGasTokenFunder Contract that funds the custom gas token.
+     */
+    constructor(
+        ArbitrumL1InboxLike _l1ArbitrumInbox,
+        ArbitrumL1ERC20GatewayLike _l1ERC20GatewayRouter,
+        address _l2RefundL2Address,
+        IERC20 _l1Usdc,
+        ITokenMessenger _cctpTokenMessenger,
+        IERC20 _customGasToken,
+        FunderInterface _customGasTokenFunder,
+        uint256 _l2MaxSubmissionCost
+    ) CircleCCTPAdapter(_l1Usdc, _cctpTokenMessenger, CircleDomainIds.Arbitrum) {
+        L1_INBOX = _l1ArbitrumInbox;
+        L1_ERC20_GATEWAY_ROUTER = _l1ERC20GatewayRouter;
+        L2_REFUND_L2_ADDRESS = _l2RefundL2Address;
+        CUSTOM_GAS_TOKEN = _customGasToken;
+        L2_MAX_SUBMISSION_COST = _l2MaxSubmissionCost;
+        CUSTOM_GAS_TOKEN_FUNDER = _customGasTokenFunder;
+        require(L1_INBOX.bridge().nativeToken() == address(_customGasToken), "Incorrect custom gas token");
```

Comment:
nit: why pass _customGasToken and check it vs just reading it from nativeToken()?